router/data_store.go

```diff
@@ -199,72 +193,69 @@ func (d *pgDataStore) List() ([]*router.Route, error) {
 	return routes, rows.Err()
 }
 
-func (d *pgDataStore) Sync(h SyncHandler, started chan<- error) {
-	idc := make(chan string)
-	if err := d.startListener(idc); err != nil {
-		started <- err
-		return
+func (d *pgDataStore) Sync(ctx context.Context, h SyncHandler, startc chan<- struct{}) error {
+	ctx, cancel := context.WithCancel(ctx)
+
+	idc, errc, err := d.startListener(ctx)
+	if err != nil {
+		return err
 	}
-	defer d.StopSync()
 
 	initialRoutes, err := d.List()
 	if err != nil {
-		started <- err
-		return
+		cancel()
```

Comment:
doesn't look like there's a need for this (or the `context.WithCancel()` above). The only thing this will do is cancel child contexts like the new `ctx` you've created above. The block below that switches on `ctx.Done()` will only execute when a parent has canceled.

Maybe just drop the `context.WithCancel()` and all uses of `cancel()` inside this method?